core/log/performance_hint.cpp

```diff
@@ -0,0 +1,190 @@
+{
+    stream << "Observed " << count << " cross-executor copies from "
+           << location_name(ptr)
+           << " that may point to unnecessary data transfers.\n";
+}
+
+
+void print_copy_to_message(std::ostream& stream, gko::uintptr ptr, int count)
+{
+    stream << "Observed " << count << " cross-executor copies to "
+           << location_name(ptr)
+           << " that may point to unnecessary data transfers.\n";
+}
+
+
+}  // namespace
+
+
+void PerformanceHint::on_allocation_completed(const Executor* exec,
+                                              const size_type& num_bytes,
+                                              const uintptr& location) const
+{
+    if (num_bytes > allocation_size_limit_) {
+        allocation_sizes_[location] = num_bytes;
+        // erase smallest allocations first
+        compact_storage(allocation_sizes_, histogram_max_size_,
+                        histogram_max_size_ * 3 / 4);
+    }
+}
+
+
+void PerformanceHint::on_free_completed(const Executor* exec,
+                                        const uintptr& location) const
+{
+    const auto it = allocation_sizes_.find(location);
+    if (it != allocation_sizes_.end()) {
+        const auto size = it->second;
+        allocation_sizes_.erase(it);
+        const auto count = ++allocation_histogram_[size];
+        if (should_log_count(count)) {
+            print_allocation_message(log(), size, count);
+        }
+        // erase rarest allocation sizes first
+        compact_storage(allocation_histogram_, histogram_max_size_,
+                        histogram_max_size_ * 3 / 4);
+    }
+}
+
+
+void PerformanceHint::on_copy_completed(const Executor* from,
+                                        const Executor* to,
+                                        const uintptr& location_from,
+                                        const uintptr& location_to,
+                                        const size_type& num_bytes) const
+{
+    if (num_bytes > copy_size_limit_ && from != to) {
+        const auto count1 = ++copy_src_histogram_[location_from];
+        const auto count2 = ++copy_dst_histogram_[location_to];
+        if (should_log_count(count1)) {
+            print_copy_from_message(log(), location_from, count1);
+        }
+        if (should_log_count(count2)) {
+            print_copy_to_message(log(), location_to, count2);
+        }
+        compact_storage(copy_src_histogram_, histogram_max_size_,
+                        histogram_max_size_ * 3 / 4);
+        compact_storage(copy_dst_histogram_, histogram_max_size_,
+                        histogram_max_size_ * 3 / 4);
+    }
+}
+
+
+void PerformanceHint::print_status() const
+{
+    for (auto entry : allocation_histogram_) {
+        if (entry.second >= 10) {
+            print_allocation_message(log(), entry.first, entry.second);
+        }
+    }
+    for (auto entry : copy_src_histogram_) {
+        if (entry.second >= 10) {
```

Comment:
Maybe it makes sense to print only for the largest count, instead of printing for every count that we track ?

Reply:
that may hide some issues, with people having to discover them step by step, instead of seeing all of them in a single spot?

Reply:
I don't follow. The histogram contains only a single count for each size, so the output of `print_status()` would be
```
[PERFORMANCE] >>> Observed 110 allocate-free pairs of size x that may point to unnecessary allocations.
```

Reply:
Okay, maybe I misunderstood something. What I mean is [something like here](https://github.com/ginkgo-project/ginkgo/pull/1035/files#diff-28c174a9777e93aa8fcb0167c2cadb033990195e8bb0edb72c98fb0d208f7ea0R62)

Reply:
That is independent of `print_status()`: The output you're seeing is immediate after the allocation took place, thus if the application has its own status output, the interspersed logger messages will allow users to see where this unnecessary allocation took place. print_status() is meant more as a tool to print a summary at the end.